Visualization for custom sources using Marker msgs

steremma opened the pull request:
# Abstract

Added a new visualization interface for all source types (plus RFID tags) to the stdr_server, #154. 
When new sources are added or deleted, an appropriate message is now published if rviz is subscribed to it (markerArrays should be enabled in Rviz).
As a result sources can be visualized not only in our custom **GUI** but also in **Rviz**.
## Implementation notes:

I created private functions able to translate the custom stdr messages into markerArray ones.
To avoid code duplication the main part of those functions is included into a template member function
responsible for creating a marker object and set every attribute that is not related to the specific source type.The specific translate functions only need to call this template and then only set the source type specific attributes themselves.

All callback functions in stdr_server responsible for adding or deleting sources used to publish towards the GUI controller. Now they also publish the (translated) message towards **rviz**, if the respective view is enabled.
## Testing:

Functionality can be tested by launching a GUI example as well as Rviz, for example:

`roslaunch stdr_launchers server_with_map_and_gui_plus_robot.launch`
`roslaunch stdr_launchers rviz.launch`

We then enable the markerArray visualization in Rviz.
Lastly we add one or more sources of any type through the stdr_gui and see it visualized in both the Gui and Rviz.
Of course adding sources in any other way (e.g through the map .yaml file) will also add visualize them in Rviz.
## Notes
1. Care has been taken to allow different source types to have the same ID.
2. Adding a source through the GUI some times gives an error message, this is also the case in the master branch
3. I am assuming that the user specifies an integer ID even though is it a string field. If that is not a case, a hash function should be added to create a unique ID for the markers.

czalidis committed:
Remove unnecessary image_loader library in cmake

Now stdr_server compiles under ROS Jade

steremma committed:
Every callback function responsible for adding sources, now also publishes towards RVIZ if it launched. As a result sources are visualized in Rviz as well as our custom GUI

Also added logic needed to remove source visualization in Rviz when a source is deleted.

steremma committed:
each source now gives a different namespace to its marker object so that different source types with the same id are allowed

pulled from hydro-devel branch

czalidis commented on stdr_server/CMakeLists.txt:
```diff
@@ -37,6 +37,7 @@ include_directories( include ${catkin_INCLUDE_DIRS})
 
 add_library(stdr_map_loader src/map_loader.cpp)
 target_link_libraries(stdr_map_loader
+    image_loader
```
This shouldn't change. It was removed at the latest commit for a reason. Please modify the commit to exclude the change, do not commit a deletion.

czalidis commented:
Thanks for the contribution! I expect to have time in order to review and test this PR, in about 2 weeks.

In the meantime some minor comments:

In my opinion the `translate*()` functions should be renamed. Actually, the word translate is confusing, because there is no translation taking place. They simply create a marker message based on the position of a source message. A name such as `*ToMarker()` (i.e. `thermalToMarker()`) would be more appropriate. They can even be templatized, based on the type of the source that they take as argument (i.e. `toMarker(thermalMsg)`).

steremma commented:
I understand that replacing those 4 functions by a template one removes
duplication but there is a problem:

The functionality depends on the message type. If all i needed was the
pose, then a template would be the perfect solution but now i also need the
I.D. The thermal, sound and co2 types all have the respective field named
"id", but the RFID message has the name "tag_id". As a result the template
should contain checks for the specific type (i.e dynamic_cast) which is
considered bad design. Furthermore, the markers color changes based on the
type. This means that out of 6 lines of code each function has, only 2 are
actually common.

GIven that fact, do you think i should just rename the functions and/or
encapsulate the pose parsing in another function, or force a template
approach however ugly it might be?

2016-03-20 0:00 GMT+02:00 Christos Zalidis notifications@github.com:

> I expect to have time in order to review and test this PR, in about 2
> weeks.
> 
> In the meantime some minor comments:
> 
> In my opinion the translate*() functions should be renamed. Actually, the
> word translate is confusing, because there is no translation taking place.
> They simply create a marker message based on the position of a source
> message. A name such as *ToMarker() (i.e. thermalToMarker()) would be
> more appropriate. They can even be templatized, based on the type of the
> source that the take as argument (i.e. toMarker(thermalMsg)).
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> https://github.com/stdr-simulator-ros-pkg/stdr_simulator/pull/184#issuecomment-198794360

steremma committed:
renamed function tranlate* to ToMarker

Similar functions dont need to have specific names, i overload them instead. This has the same effect as a function template with specialization would, but also enchances readibility (imo)

trainman419 commented:
Sounds like a good use case for explicit template specialization.

czalidis commented on stdr_server/src/stdr_server.cpp:
```diff
@@ -83,7 +83,11 @@ namespace stdr_server {
+    
+    //!< Rviz publisher
+    _sourceVectorPublisherRviz = _nh.advertise<visualization_msgs::MarkerArray>(
+      "/visualization_marker_array", 1, true);
```
Rename the topic to `stdr_server/sources_visualization_markers`.

czalidis commented on stdr_server/src/stdr_server.cpp:
```diff
@@ -159,11 +163,19 @@ namespace stdr_server {
+      RFIDMarkerArray.markers.push_back(toMarker(it->second)); 
     }
     _rfidTagVectorPublisher.publish(rfidTagList);
+    if (_sourceVectorPublisherRviz.getNumSubscribers() >= 1)
```
Although I see the reason why you don't want to publish if no one is listening, this contradicts the use of a latched publisher. Declaring the publisher as latched (which is the right thing to do here) means that whenever a new subscriber "listens" to the markers topic, will get the last message. When this subscriber is the Rviz, means that even if you launch Rviz well after you insert a new source, the sources will get drawn. So I suggest removing the if-statements preventing the publication of marker message for every source (for all types of sources).

czalidis commented on stdr_server/src/stdr_server.cpp:
```diff
@@ -682,5 +752,126 @@ namespace stdr_server {
+  {
+    visualization_msgs::Marker marker;
+    // Set the frame ID and timestamp.  See the TF tutorials for information on these.
+    marker.header.frame_id = "/world";
```
The reference frame_id should be `/map_static`, as this is the "world" frame in terms of simulation. Otherwise the map origin is not taken into account.

czalidis commented on stdr_server/src/stdr_server.cpp:
```diff
@@ -682,5 +752,126 @@ namespace stdr_server {
+    visualization_msgs::Marker marker;
+    // Set the frame ID and timestamp.  See the TF tutorials for information on these.
+    marker.header.frame_id = "/world";
+    marker.header.stamp = ros::Time::now();
```
You should use `ros::Time()` instead, making the marker independent of the creation time, when it comes to drawing.

czalidis commented:
Other than the inline comments this looks good! Thanks!

Please consider squashing your commits or creating a new PR with a cleaner commit tree.

steremma committed:
changed marker topic to a more descriptive name. All sources now publish their status even if RVIZ is not yet launched. Fixed marker elements - frame id and timestamp

steremma commented:
I included all changes proposed after code review. Squashed unnecessary commits to present a clear commit tree. Change my author and commiter name for all commits.

steremma committed:
Sources are now properly deleted on Rviz.

steremma commented:
Fixed source deletion of Rviz.

czalidis commented:
One issue still remains: Since all source types use the same publisher for markers, subsequent messages for a specific source type overwrite the visualization state for all other sources.

Better demonstrated with an example:
Launch STDR GUI and add 2 thermal sources and one CO2. Then delete the CO2 source. Launch Rviz and visualize the markers (or echo the markers topic). No markers come up (empty message). The last message for the CO2 source, overwrites the markers for the other sources.

I can not think of an elegant solution for this issue, other that using separate publishers for different types of sources (which I don't like that much) or publishing a marker message after queering all sources about their current state.

@steremma what do you think? Any other solution? 

Keep in mind that we can ignore this issue if we agree that this is a rare use case and does not worth dealing with it, but personally I think that this behavior (wrong visualization) is a bit confusing for the user.

steremma committed:
Now republishing existing sources to RVIZ after every successful deletion.

steremma commented:
This bug can only be reproduced when both source additions and the source deletion take place 
<b> before </b> the user launches Rviz. If Rviz is launched simultaneously with (or before) the simulator (which i consider the common case) everything works correctly.

However there is a simple way to solve the issue however rare it might be, since the information about existing sources is not lost, i.e <b>CO2SourceList</b> still contains all previous sources even if they are not visualized. This means that all we have to do is republish the existing sources to Rviz after every successful deletion.
I wrapped this up into a private function, called after every deletion. It seems to be passing all tests, including the previously failed one. It would be interesting if we could automate this testing for future use too.

czalidis commented:
I have added the republication of markers also for the addition of sources and properly rebased your commits in a new branch: [steremma-markers](https://github.com/stdr-simulator-ros-pkg/stdr_simulator/tree/steremma-markers).

I have tested the changes for various use cases, please test them yourself to verify that all issues have been resolved before merging.

czalidis commented:
Merged in 8f64bf3
